src/concurrency.h

```diff
@@ -8,140 +8,178 @@
-    {}
+    explicit single_consumer_queue< T >( unsigned int cap = QUEUE_MAX_SIZE,
+                                         std::function< void( T const & ) > on_drop_callback = nullptr )
+        : _cap( cap )
+        , _accepting( true )
+        , _on_drop_callback( on_drop_callback )
+    {
+    }
 
+    // Enqueue an item onto the queue.
+    // If the queue grows beyond capacity, the front will be removed, losing whatever was there!
     void enqueue(T&& item)
     {
         std::unique_lock<std::mutex> lock(_mutex);
-        if (_accepting)
+        if( ! _accepting )
         {
-            _queue.push_back(std::move(item));
-            if (_queue.size() > _cap)
-            {
-                if (_on_drop_callback)
-                {
-                    _on_drop_callback(_queue.front());
-                }
-                _queue.pop_front();
-            }
+            if( _on_drop_callback )
+                _on_drop_callback( item );
+            return;
         }
+
+        _queue.push_back(std::move(item));
+
+        if( _queue.size() > _cap )
+        {
+            if( _on_drop_callback )
+                _on_drop_callback( _queue.front() );
+            _queue.pop_front();
+        }
+
         lock.unlock();
+
+        // We pushed something -- let others know there's something to dequeue
         _deq_cv.notify_one();
     }
 
-    void blocking_enqueue(T&& item)
-    {
-        auto pred = [this]()->bool { return _queue.size() < _cap || _need_to_flush; };
 
+    // Enqueue an item, but wait for room if there isn't any
+    // Returns true if the enqueue succeeded
+    bool blocking_enqueue(T&& item)
+    {
         std::unique_lock<std::mutex> lock(_mutex);
-        if (_accepting)
+        _enq_cv.wait( lock, [this]() {
+            return _queue.size() < _cap; } );
+        if( ! _accepting )
         {
-            _enq_cv.wait(lock, pred);
-            _queue.push_back(std::move(item));
+            // We shouldn't be adding anything to the queue when we're stopping
+            if( _on_drop_callback )
+                _on_drop_callback( item );
+            return false;
         }
+
+        _queue.push_back(std::move(item));
         lock.unlock();
+
+        // We pushed something -- let another know there's something to dequeue
         _deq_cv.notify_one();
+
+        return true;
     }
 
 
-    bool dequeue(T* item ,unsigned int timeout_ms)
+    // Remove one item; if unavailable, wait for it
+    // Return true if an item was removed -- otherwise, false
+    bool dequeue( T * item, unsigned int timeout_ms )
     {
         std::unique_lock<std::mutex> lock(_mutex);
-        _accepting = true;
-        _was_flushed = false;
-        const auto ready = [this]() { return (_queue.size() > 0) || _need_to_flush; };
-        if (!ready() && !_deq_cv.wait_for(lock, std::chrono::milliseconds(timeout_ms), ready))
+        if( ! _deq_cv.wait_for( lock,
```

Comment:
This if statement is really confusing..
too much "!" operators..
I suggest rewrite it as a positive if so it will be more understandable.

Reply:
Everything with these predicates is confusing :)